Backend/app/rider-platform/rider-app/Main/src/SharedLogic/DirectionsCache.hs

```diff
@@ -0,0 +1,108 @@
+getRoutes merchantId req = do
+  merchant <- QMerchant.findById merchantId >>= fromMaybeM (MerchantNotFound merchantId.getId)
+  let origin = NE.head req.waypoints
+  let dest = NE.last req.waypoints
+  originGeoHash <- case DG.encode merchant.geoHashPrecisionValue (origin.lat, origin.lon) of
+    Just hash -> return $ pack hash
+    Nothing -> throwError $ InternalError "Failed to compute Origin GeoHash" -- This case will never happen as we are getting lat long from Google and hence a valid geo hash will always be possible for a valid lat long.
+  destGeoHash <- case DG.encode merchant.geoHashPrecisionValue (dest.lat, dest.lon) of
+    Just hash -> return $ pack hash
+    Nothing -> throwError $ InternalError "Failed to compute Destination GeoHash" -- This case will never happen as we are getting lat long from Google and hence a valid geo hash will always be possible for a valid lat long.
+  timeSlot <- getSlot merchantId
+  case timeSlot of
+    Just tmeSlt ->
+      do
+        cachedResp <- DQ.findRoute originGeoHash destGeoHash (pack tmeSlt)
+        case cachedResp of
+          Just resp -> convertFromDirCache resp
+          Nothing -> callDirectionsApi merchantId req originGeoHash destGeoHash tmeSlt
+    Nothing ->
+      Maps.getRoutes merchantId req
+
+callDirectionsApi :: (EncFlow m r, CacheFlow m r, EsqDBFlow m r, CoreMetrics m) => Id Merchant.Merchant -> Maps.GetRoutesReq -> Text -> Text -> String -> m Maps.GetRoutesResp
+callDirectionsApi merchantId req originGeoHash destGeoHash timeSlot = do
+  resp <- Maps.getRoutes merchantId req
+  if Kernel.Prelude.null resp
+    then throwError $ InternalError "Null response from Directions API" -- This case will never occure unless Google's Direction API Fails.
+    else do
+      let (cachedResp : _) = resp
+      directionsCache <- convertToDirCache originGeoHash destGeoHash (pack timeSlot) cachedResp
+      Esq.runTransaction $ DQ.create directionsCache
+      DCC.cacheDirectionsResponse directionsCache
+      return resp
+
+getSlot :: (MonadIO m, CacheFlow m r, EsqDBFlow m r, EsqDBReplicaFlow m r, EncFlow m r) => Id Merchant.Merchant -> m (Maybe String)
+getSlot merchantId = do
+  utcTime <- liftIO DT.getCurrentTime
+  let istTime = utcToLocalTime (TimeZone 330 False "IST") utcTime
+  let timeOfDay = localTimeOfDay istTime
+  slots <- getSlots merchantId
+  return $ matchSlot timeOfDay slots
+
+getSlots :: (CacheFlow m r, EsqDBFlow m r, EsqDBReplicaFlow m r, EncFlow m r) => Id Merchant.Merchant -> m [Slot]
+getSlots merchantId = do
+  configs <- M.findById merchantId
+  case configs of
+    Just merchant -> return merchant.dirCacheSlot
+    Nothing -> throwError $ InternalError "Error in fetching configs from Database"
+
+matchSlot :: TimeOfDay -> [Slot] -> Maybe String
+matchSlot _ [] = Nothing
+matchSlot currTime (slot : slots)
+  | slot.startTime <= currTime && currTime < slot.endTime = Just $ unpack slot.slot
```

Comment:
How are you handling time before 12AM and after 12AM?

Reply:
The time from 20:00 to 23:59:59 and 00:00 to 8:00 both are S4

Reply:
How this logic is going to satisfy request coming at 23:00:00? Can you explain?